### ISSUE TITLE
What are the expected packet contents for packets with QEO_DECRYPTION_FAILED?

### ISSUE THREAD

**Issue by stevedoyle** [microsoft/quic-offloads#32]
For received packets that fail decryption and have a status if QEO_DECRYPTION_FAILED, is there any expectation on the contents of the packet? For example, the packet could be the original, still encrypted packet, that was received, or it could be the contents produced by the failed decryption. 

From a HW perspective, providing the contents produced by the failed decryption is probably simpler.

**Comment by nibanks** [comment]
Yeah, it makes sense that the resulting payload of a failed attempt to decrypt would be modified. I was originally thinking we needed to pass up this failed payload for stateless reset scenarios (on client), but I've since realized / been reminded that, at least in the case where the client uses a CID, the CID won't match so decryption wouldn't even be attempted.

So, the only case we'd want the payload passed up would be if the client doesn't use a CID, and the received packet is a stateless reset. In this case we'd still need at least the original 16 bytes of what could be the stateless reset token passed up. But since the stateless reset token is the end of the packet, which is where the authentication tag is, I think it should be unmodified.

But for servers, I don't even think we need (want?) failed decryptions to be passed up. It could actually be a really nice DDoS defense (from random packets) to have the HW drop the packets that fail to decrypt. But then the HW likely would have to inform the OS (at least periodically) of a counter of failed decryptions, right?

Thoughts?

cc @borispis

**Comment by stevedoyle** [comment]
> But for servers, I don't even think we need (want?) failed decryptions to be passed up. It could actually be a really nice DDoS defense (from random packets) to have the HW drop the packets that fail to decrypt. But then the HW likely would have to inform the OS (at least periodically) of a counter of failed decryptions, right?

Yes, I agree. Dropping in HW would be a good DDoS defense and can even simplify the HW flow but as you say it does require the HW to maintain counters of failed decryptions, per key phase. If the HW maintains these counters should the SW/OS periodically query them or should HW send notifications (time based or threshold based)? Are there other counters that the HW should track?

**Comment by BorisPis** [comment]
> Yeah, it makes sense that the resulting payload of a failed attempt to decrypt would be modified. I was originally thinking we needed to pass up this failed payload for stateless reset scenarios (on client), but I've since realized / been reminded that, at least in the case where the client uses a CID, the CID won't match so decryption wouldn't even be attempted.
> 
> So, the only case we'd want the payload passed up would be if the client doesn't use a CID, and the received packet is a stateless reset. In this case we'd still need at least the original 16 bytes of what could be the stateless reset token passed up. But since the stateless reset token is the end of the packet, which is where the authentication tag is, I think it should be unmodified.
> 
> But for servers, I don't even think we need (want?) failed decryptions to be passed up. It could actually be a really nice DDoS defense (from random packets) to have the HW drop the packets that fail to decrypt. But then the HW likely would have to inform the OS (at least periodically) of a counter of failed decryptions, right?
> 
> Thoughts?
> 
> cc @BorisPis

Yes, that makes sense, dropping and counting seems good to me. Perhaps, for debugging purposes, passing packets with failed authentication can still be useful.
